File: src/main/java/bad/robot/radiate/Progress.java
package bad.robot.radiate;

public class Progress {

    private int current;
    private int numberOfBuilds = 1;
    private final double max;

    public Progress(int current, int max) {
        this.current = current;
        this.max = max;
    }

    private Progress(int current, int max, int numberOfBuilds) {
        this(current, max);
        this.numberOfBuilds = numberOfBuilds;
    }

    public void increment() {
        current++;
    }

    public void decrement() {
        current--;
    }

    public Progress add(Progress progress) {
        return new Progress(current + progress.current, (int) (max + progress.max), numberOfBuilds(progress));
    }

    private int numberOfBuilds(Progress progress) {
        return numberOfBuilds() + progress.numberOfBuilds();
    }

    public int asAngle() {
        return - asPercentageOf(360).intValue();
    }

    private int asPercentage() {
        return asPercentageOf(100).intValue();
    }

    private Double asPercentageOf(int i) {
        return current / max * i;
    }

    public boolean complete() {
        return current >= max;
    }

    public int numberOfBuilds() {
        return numberOfBuilds;
    }

    @Override
    public String toString() {
        return asPercentage() + "%";
    }

    public boolean lessThan(Progress progress) {
        return asPercentage() < progress.asPercentage();
    }

    public boolean greaterThan(Progress progress) {
        return asPercentage() > progress.asPercentage();
    }
}

File: src/main/scala/bad/robot/radiate/Progress.scala
package bad.robot.radiate

case class Progress(private var current: Int, private var max: Double) {

  private var _numberOfBuilds = 1

  private def this(current: Int, max: Int, numberOfBuilds: Int) = {
    this(current, max)
    this._numberOfBuilds = numberOfBuilds
  }

  def increment() {
    current += 1
  }

  def decrement() {
    current -= 1
  }

  def add(progress: Progress): Progress = {
    new Progress(current + progress.current, (max + progress.max).toInt, numberOfBuilds(progress))
  }

  private def numberOfBuilds(progress: Progress): Int = {
    numberOfBuilds + progress.numberOfBuilds
  }

  def asAngle: Int = {
    -asPercentageOf(360).intValue
  }

  private def asPercentage: Int = {
    asPercentageOf(100).intValue
  }

  private def asPercentageOf(i: Int): Double = {
    current / max * i
  }

  def complete: Boolean = {
    current >= max
  }

  def numberOfBuilds: Int = {
    _numberOfBuilds
  }

  def <(progress: Progress): Boolean = {
    asPercentage < progress.asPercentage
  }

  def >(progress: Progress): Boolean = {
    asPercentage > progress.asPercentage
  }

  override def toString: String = {
    asPercentage + "%"
  }
}
